M2/Macaulay2/d/CMakeLists.txt

```diff
@@ -0,0 +1,199 @@
+###############################################################################
+## This directory contains the rules for building M2-interpreter
+
+include(scc) ## CMake macro for calling the Safe C Compiler
+
+# TODO: changing this causes a lot of recompiling
+## Configure distributed-packages.h at the destination
+string(REPLACE ";" " " DISTRIBUTED_PACKAGES "${DISTRIBUTED_PACKAGES}")
+# TODO: configure this header directly in version.dd?
+file(GENERATE OUTPUT distributed-packages.h CONTENT "#define DISTRIBUTED_PACKAGES \"${DISTRIBUTED_PACKAGES}\"")
+
+## Configure version.dd at the destination
+#configure_file(version.dd.in version.dd #ONLY)
+
+###############################################################################
+
+set(CLIST
+  M2types.c M2lib.c M2mem.c types.h
+  scclib.c ../c/scc-core.h
+  memdebug.c memdebug.h
+  debug.c debug.h
+  distributed-packages.h
+  getpagesize.h M2mem.h
+  gmp_aux.c gmp_aux.h # TODO: needed with mpir?
+  gdbm_interface.c
+  )
+
+# CAUTION: order is important in this list
+# TODO: can there be #ifdefs in D sources?
+set(DLIST
```

Comment:
Could we arrange for lists like this to be shared by both build systems?  That would prevent certain sorts of future errors.

Reply:
The way the `d` directory is currently built, I don't know of a clean way to do this, because building some `.d` sources depends on flags (e.g. XML, Python, or MySQL). However, if `scc1` supported `#ifdef`s in its syntax, then we could put the list in a file and build everything in it, knowing that the compiler would skip the content if a flag is not set.

Reply:
For that reason, you've convinced me that it would be good to preprocess the D files.  The way to do that is this: make a shell script called `c/scc`.  It would run `cpp` on the file first and then run `c/scc1`.  The code of `scc1` would be modified to pay attention to the lines left behind `cpp` : a line such as `# 11 "foo.d" 2` would tell it that the next line (check that) is line 11 from the file "foo.d".  We need that information for giving proper line numbers in error messages, and we don't want to be off even by 1.

The reason the "1" is in the name "scc1" is to allow for this, I've just remembered.

Feel free to make this comment thread into an issue.

Reply:
Made it an issue in #1197. Setting this thread as resolved for now.

Reply:
@mikestillman  may like to do this for the `e` directory.  If there are no conditionals, how would that be done?  (Not saying to do it now ... )

Reply:
This would be a little complicated right now, since there are multiple classes of sources. Perhaps after some engine refactoring it'll be nicer.